sklearn/semi_supervised/tests/test_label_propagation.py

```diff
@@ -29,69 +31,69 @@
-            )
-        else:
-            assert_array_almost_equal(
-                np.asarray(clf.label_distributions_[2]), np.array([0.5, 0.5]), 2
-            )
-
-
-def test_predict():
-    samples = [[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]]
+    clf = Estimator(**parameters).fit(samples, labels)
+    # unstable test; changes in k-NN ordering break it
+    if parameters["kernel"] != "knn":
+        assert_allclose(
+            np.asarray(clf.label_distributions_[2]), np.array([0.5, 0.5]), 2
+        )
+
+
+@pytest.mark.parametrize("Estimator, parameters", ESTIMATORS)
+def test_predict(global_dtype, Estimator, parameters):
+    samples = np.asarray([[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]], dtype=global_dtype)
     labels = [0, 1, -1]
-    for estimator, parameters in ESTIMATORS:
-        clf = estimator(**parameters).fit(samples, labels)
-        assert_array_equal(clf.predict([[0.5, 2.5]]), np.array([1]))
+    clf = Estimator(**parameters).fit(samples, labels)
+    assert_array_equal(clf.predict([[0.5, 2.5]]), np.array([1]))
 
 
-def test_predict_proba():
-    samples = [[1.0, 0.0], [0.0, 1.0], [1.0, 2.5]]
+@pytest.mark.parametrize("Estimator, parameters", ESTIMATORS)
+def test_predict_proba(global_dtype, Estimator, parameters):
+    samples = np.asarray([[1.0, 0.0], [0.0, 1.0], [1.0, 2.5]], dtype=global_dtype)
     labels = [0, 1, -1]
-    for estimator, parameters in ESTIMATORS:
-        clf = estimator(**parameters).fit(samples, labels)
-        assert_array_almost_equal(
-            clf.predict_proba([[1.0, 1.0]]), np.array([[0.5, 0.5]])
-        )
+    clf = Estimator(**parameters).fit(samples, labels)
+    assert_allclose(clf.predict_proba([[1.0, 1.0]]), np.array([[0.5, 0.5]]))
 
 
-def test_label_spreading_closed_form():
+@pytest.mark.parametrize("alpha", [0.1, 0.3, 0.5, 0.7, 0.9])
+@pytest.mark.parametrize("Estimator, parameters", ESTIMATORS)
+def test_label_spreading_closed_form(global_dtype, Estimator, parameters, alpha):
     n_classes = 2
     X, y = make_classification(n_classes=n_classes, n_samples=200, random_state=0)
+    X = X.astype(global_dtype, copy=False)
     y[::3] = -1
     clf = label_propagation.LabelSpreading().fit(X, y)
     # adopting notation from Zhou et al (2004):
     S = clf._build_graph()
     Y = np.zeros((len(y), n_classes + 1))
     Y[np.arange(len(y)), y] = 1
     Y = Y[:, :-1]
-    for alpha in [0.1, 0.3, 0.5, 0.7, 0.9]:
-        expected = np.dot(np.linalg.inv(np.eye(len(S)) - alpha * S), Y)
-        expected /= expected.sum(axis=1)[:, np.newaxis]
-        clf = label_propagation.LabelSpreading(max_iter=10000, alpha=alpha)
-        clf.fit(X, y)
-        assert_array_almost_equal(expected, clf.label_distributions_, 4)
+    expected = np.dot(np.linalg.inv(np.eye(len(S)) - alpha * S), Y)
+    expected /= expected.sum(axis=1)[:, np.newaxis]
+    # Inverting might cause zeros values to become NaN values in the case of float32.
+    # Hence, we replace those NaN values to 0.
```

Comment:
That's strange because we divide by expected.sum(axis=1) which means that a whole row is full 0s. Probably worth checking if everything is as expected until this point

Reply:
I'd rather use ``np.multiply(A, B, where=...)`` to avoid the division by zero warning (and better practice imo)

Reply:
Thanks!